### PR TITLE
Remove upcast to float64 in the graph.

### PR DESCRIPTION
**Pull request by nouiz**
finish gh-4438.

### CODE REVIEW

File: theano/tensor/opt.py [hunk 1]
```diff
@@ -5210,11 +5210,12 @@ def local_opt_alloc(node):
                     val = get_scalar_constant_value(input)
                     assert val.size == 1
                     # check which type of op
+                    casted = T.mul(*shapes).astype(str(val.dtype))
```
Comment (lamblin) [inline theano/tensor/opt.py]: Wait, `val.dtype` does not seem to be the right dtype here, I think we should use `node.outputs[0].dtype`.

File: theano/tensor/opt.py [hunk 2]
```diff
@@ -5226,11 +5227,12 @@ def local_opt_alloc(node):
                     to_prod = [shapes[i] for i in xrange(len(shapes))
                                if i in node.op.axis]
                     if to_prod:
+                        casted = T.mul(*shapes).astype(str(val.dtype))
```
Comment (lamblin) [inline theano/tensor/opt.py]: Same here.

File: theano/tensor/opt.py (hunk 1 above)
Comment (nouiz) [inline theano/tensor/opt.py]: Why? node.outputs[0].dtype should be val.dtype. Alloc output dtype is val dtype. Alloc just broadcast a value, it do not change the type.

Comment (lamblin) [inline theano/tensor/opt.py]: `get_scalar_constant_value` digs through `cast` operations.
You could use `input.dtype`, though.

Comment (nouiz) [inline theano/tensor/opt.py]: done

**Comment by lamblin** [comment]
OK to merge when Travis passes.

**Comment by cesarsouza** [comment]
May I ask if this issue and its fix were related only to the multiplication operation? I am adding two float32 tensors together (one that comes from a quite complex set of computations) and their result is also being upcast to float64. I am using Theano from the master branch, cloned about two days ago.

**Comment by lamblin** [comment]
It was related to the `sum` and `prod` reduction operations, it should not affect the elemwise addition or multiplication.

**Pull request merged**